transient_current_dave_plugin/dave_world_plugins/src/UnderwaterCurrentPlugin.cc

```diff
@@ -0,0 +1,358 @@
+    << std::endl;
+  this->currentHorzAngleModel.Print();
+
+  // Read the depth dependent ocean current file path from the SDF file
+  if (currentVelocityParams->HasElement("database"))
+    this->databaseFilePath =
+      currentVelocityParams->Get<std::string>("database");
+  else
+    this->databaseFilePath = "world://transientOceanCurrentDatabase.csv";
+
+  GZ_ASSERT(!this->databaseFilePath.empty(),
+    "Empty database file path");
+
+  // Read database
+  std::ifstream csvFile; std::string line;
+  csvFile.open(this->databaseFilePath);
+  // skip the 3 lines
+  getline(csvFile, line); getline(csvFile, line); getline(csvFile, line);
+  while (getline(csvFile, line))
+  {
+      if (line.empty())  // skip empty lines:
+      {
+          continue;
+      }
+      std::istringstream iss(line);
+      std::string lineStream;
+      std::string::size_type sz;
+      std::vector <long double> row;
+      while (getline(iss, lineStream, ','))
+      {
+          row.push_back(stold(lineStream, &sz));  // convert to double
+      }
+      ignition::math::Vector3d read;
+      read.X() = row[0]; read.Y() = row[1]; read.Z() = row[2];
+      this->database.push_back(read);
+  }
+
+  // Initialize the time update
+#if GAZEBO_MAJOR_VERSION >= 8
+  this->lastUpdate = this->world->SimTime();
+#else
+  this->lastUpdate = this->world->GetSimTime();
+#endif
+  this->currentVelModel.lastUpdate = this->lastUpdate.Double();
+  this->currentHorzAngleModel.lastUpdate = this->lastUpdate.Double();
+  this->currentVertAngleModel.lastUpdate = this->lastUpdate.Double();
+
+  // Advertise the current velocity topic
+  this->publishers[this->currentVelocityTopic] =
+    this->node->Advertise<msgs::Vector3d>(
+    this->ns + "/" + this->currentVelocityTopic);
+
+  gzmsg << "Current velocity topic name: " <<
+    this->ns + "/" + this->currentVelocityTopic << std::endl;
+
+  // Subscribe vehicle depth topic
+  this->vehicleDepthTopic = "vehicle_depth";
+  this->subscriber = this->node->Subscribe<msgs::Any>(
+    this->vehicleDepthTopic, &UnderwaterCurrentPlugin::SubscribeVehicleDepth,
+    this);
+
+  // Connect the update event
+  this->updateConnection = event::Events::ConnectWorldUpdateBegin(
+    boost::bind(&UnderwaterCurrentPlugin::Update,
+    this, _1));
+
+  gzmsg << "Underwater current plugin loaded!" << std::endl
+    << "\tWARNING: Current velocity calculated in the ENU frame"
+    << std::endl;
+}
+
+/////////////////////////////////////////////////
+void UnderwaterCurrentPlugin::Init()
+{
+  // Doing nothing for now
+}
+
+/////////////////////////////////////////////////
+void UnderwaterCurrentPlugin::Update(const common::UpdateInfo & /** _info */)
+{
+#if GAZEBO_MAJOR_VERSION >= 8
+  common::Time time = this->world->SimTime();
+#else
+  common::Time time = this->world->GetSimTime();
+#endif
+  // Calculate the flow velocity and the direction using the Gauss-Markov
+  // model
+
+  // Update current velocity
```

Comment:
How would you feel about making the current generation take place in a separate ROS node that has a `ServiceServer` interface. This would move the database loading and reading from this plugin into the ROS node.

The benefits of this approach would be to hide implementation details of how the current is calculated for a given location and depth. Other groups could then update that node, or write their own custom nodes that implement the same interface. This would allow for calculating the current based on physical models, or for looking up databases containing ocean currents in locations all around the world.

Some examples of ocean currents databases:

[World Ocean Circulation Experiment](https://www.nodc.noaa.gov/woce/)
[Global Ocean Currents Database](https://www.ncei.noaa.gov/access/data/global-ocean-currents-database/)
[Ocean Surface Current Analyses Real-time](https://www.esr.org/research/oscar/)

A possible service definition for the interface is

```
std_msgs/Time time  # Time of desired current, 0 is for most recent time
geographic_msgs/GeoPose geo_pose  # Pose where desired currents are located, typically a vehicle location
---
geometry_msgs/Twist twist  # Velocity of current
```

If you are interested in this I can work on providing a pull request.

Reply:
Thank you very much for the comments and links. The team is having continuous conversations over this topic. I believe that this will be dealt with sometime near future. Regrettably, the current PR is to provide initial basic functionality. We will keep track of the idea!